### DISCUSSION TITLE
What is suppose to happen when aircraft reach 90 degree of pitch

### DISCUSSION THREAD

**Discussion by ShirelJosef** [JSBSim-Team/jsbsim#924] [Q&A]
I am initializing aircraft with 0,0,0 pitch roll yaw and gives it 1 in throttle and -1 in elevator.
the aircraft pitch rise until it reach 90 and then it changes continuously making a very fast non physical roll (I visualize it using opengl).
Is it expected?
Although from the graph it seems immediate, the change is not immediate
gifs:
![output2](https://github.com/JSBSim-Team/jsbsim/assets/26959512/28c4f589-b2cc-4d28-a1e8-3c1cb086dfa4)
![output](https://github.com/JSBSim-Team/jsbsim/assets/26959512/48bfbf1d-15f1-4de9-acac-89f33f1f5929)
![Figure_1](https://github.com/JSBSim-Team/jsbsim/assets/26959512/3a4e0882-0839-47a2-9a3f-deb0f9802675)

```python 
import jsbsim

import numpy as np
import matplotlib.pyplot as plt

def print_values(simulation):
    print("attitude/phi-deg "+str( simulation.get_property_value("attitude/phi-deg")))
    print("attitude/theta-deg "+str( simulation.get_property_value("attitude/theta-deg")))
    print("attitude/psi-deg "+str( simulation.get_property_value("attitude/psi-deg")))
    print("position/h-sl-ft "+str( simulation.get_property_value("position/h-sl-ft")))
    print("\n")

roll_list = []
pitch_list = []
yaw_list = []

# Short example script to create and run the environment with
# constant action for 1 simulation second.
if __name__ == "__main__":
    from time import sleep
    root='.'
    # Initialize JSBSim
    simulation = jsbsim.FGFDMExec(root, None)
    simulation.set_debug_level(0)
    simulation.set_dt(0.01)

    initial_long_deg = 0
    intial_lat_deg = 0
    # Load F-16 model and set initial conditions
    simulation.load_model('f16')
    simulation.set_property_value('propulsion/set-running', -1)
    simulation.set_property_value('ic/h-sl-ft', 10000)
    simulation.set_property_value('ic/long-gc-deg', initial_long_deg)
    simulation.set_property_value('ic/lat-geod-deg', intial_lat_deg)
    simulation.set_property_value('ic/u-fps', 1000.)

    simulation.run_ic()
    print_values(simulation)
    
    t = np.arange(6000)
    for i in t:
        simulation.set_property_value("propulsion/tank/contents-lbs", 1000)
        simulation.set_property_value("propulsion/tank[1]/contents-lbs", 1000)

        # Set gear up
        simulation.set_property_value("gear/gear-cmd-norm", 0.0)
        simulation.set_property_value("gear/gear-pos-norm", 0.0)
        simulation.set_property_value("fcs/aileron-cmd-norm", 0)
        simulation.set_property_value("fcs/elevator-cmd-norm", -1)
        simulation.set_property_value("fcs/rudder-cmd-norm", 0)
        simulation.set_property_value("fcs/throttle-cmd-norm", 1)
        simulation.run()

        pitch = simulation.get_property_value("attitude/pitch-rad")
        roll = simulation.get_property_value("attitude/roll-rad")
        yaw =  simulation.get_property_value("attitude/heading-true-rad")

        roll_list.append(roll)
        pitch_list.append(pitch)
        yaw_list.append(yaw)
        '''if pitch > np.pi/2 - 0.01:
            print("break")
            print("pitch"+str((pitch)))
            print("roll"+str((roll)))
            print("yaw"+str((yaw)))
            print("\n")'''

    fig, axs = plt.subplots(3 ,sharex=True)
    axs[0].plot(t, roll_list)
    axs[1].plot(t, pitch_list)
    axs[2].plot(t, yaw_list)
    axs[0].set_title('roll')
    axs[1].set_title('pitch')
    axs[2].set_title('yaw')
    plt.show()
```

**Comment by seanmcleod** [comment]
Hmm, so the roll angle toggles between 0 and pi when passing through +-90 degrees of pitch.

Possibly due to 'gimbal lock' during conversion from quaternions to Euler angles.

https://github.com/JSBSim-Team/jsbsim/blob/6c1abbe2ba53268eaf2178bebd946af0e49c7d43/src/math/FGMatrix33.cpp#L156-L190

**Comment by seanmcleod** [comment]
So just double-checking, I logged the roll rate `velocities/p-rad_sec`, displayed as deg/s in the graph.

![image](https://github.com/JSBSim-Team/jsbsim/assets/824856/6ed49e33-cc94-4af0-8867-962e98dae3b5)

So with a roll rate so small there is definitely no way that the aircraft is actually rolling through 180 degrees in a couple of simulation frames. So I'm pretty sure the issue is related to converting between quaternions and Euler angles.

Notice the same issue with the yaw, also switches virtually instantly between 0/360 (north) and 180 (south) each time the pitch attitude passes through +-90 degrees.

**Reply by ShirelJosef** [reply]
Yes, this is an issue in the conversion
I used to following code:
``` python
from scipy.spatial.transform import Rotation as R
m_rot = R.from_matrix(propagate.get_Tl2b().T)
f16_roll, f16_pitch, f16_yaw = r = m_rot.as_euler('zyx')
```
And I get smooth transition.
**Isn't this a bug?**

Also, I need to use angular velocities in the body frame ( p,q,r I assume), but now I think it might suffer from similar issues. 
![Figure_1-1](https://github.com/JSBSim-Team/jsbsim/assets/26959512/fa7a6147-07a2-434a-93e5-f9776b890734)

I tried to get the angular velocities from computing relative rotations with quaternions from the previous and current R.from_matrix(propagate.get_Tl2b().T) but it doesn't seem correct
(red is from jsbsim)
![Figure_1-2](https://github.com/JSBSim-Team/jsbsim/assets/26959512/4ae207ed-41c2-423a-be16-fd092ec29a8b)
using this code from https://mariogc.com/post/angular-velocity-quaternions/
```python 
def angular_velocities(q1, q2, dt):
    return (2 / dt) * np.array([
        q1[0]*q2[1] - q1[1]*q2[0] - q1[2]*q2[3] + q1[3]*q2[2],
        q1[0]*q2[2] + q1[1]*q2[3] - q1[2]*q2[0] - q1[3]*q2[1],
        q1[0]*q2[3] - q1[1]*q2[2] + q1[2]*q2[1] - q1[3]*q2[0]])
```

**Comment by seanmcleod** [comment]
> And I get smooth transition.

So what does your graph output look like now? 

> Notice the same issue with the yaw,....

Ignore that comment, yaw should be switching between north and south at the top and bottom of the loop.

**Reply by ShirelJosef** [reply]
![Screenshot from 2023-06-11 18-00-31](https://github.com/JSBSim-Team/jsbsim/assets/26959512/08ccf9be-8111-44e7-8e1d-eb6f0c5bc1db)

the roll and psi cancel each other near gimbal lock values for example:
before reaching 90 pitch
f16_roll -0.19745271797018438
f16_pitch 1.5658956076812902
f16_yaw 0.19770115798908572
after reaching 90 pitch
f16_roll -3.038253346978119
f16_pitch 1.561270932793192
f16_yaw 3.0385035131695073

**Comment by seanmcleod** [comment]
In terms of the yaw, as shown in the graphs above do you agree that the output is correct? Just take into account the wrap-around 0/360 deg.

**Comment by agodemar** [comment]
It's a well-known problem. The moment you use Euler angles for visualization you get this. If you use a visualization function that directly handles the aircraft attitude quaternion, the motion should be correctly displayed
